rubullet/src/types.rs

```diff
@@ -0,0 +1,1128 @@
+
+/// The unique ID for a Visual Shape
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct VisualId(pub c_int);
+
+/// The unique ID for a Collision Shape
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct CollisionId(pub c_int);
+
+/// The unique ID for a Texture
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct TextureId(pub(crate) c_int);
+
+/// The unique ID for a User Debug Parameter Item
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct ItemId(pub(crate) c_int);
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+/// An enum to represent different types of joints
+pub enum JointType {
+    Revolute = 0,
+    Prismatic = 1,
+    Spherical = 2,
+    Planar = 3,
+    Fixed = 4,
+    Point2Point = 5,
+    Gear = 6,
+}
+
+impl TryFrom<i32> for JointType {
+    type Error = Error;
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(JointType::Revolute),
+            1 => Ok(JointType::Prismatic),
+            2 => Ok(JointType::Spherical),
+            3 => Ok(JointType::Planar),
+            4 => Ok(JointType::Fixed),
+            5 => Ok(JointType::Point2Point),
+            6 => Ok(JointType::Gear),
+            _ => Err(Error::new("could not convert into a valid joint type")),
+        }
+    }
+}
+#[derive(Debug)]
+/// Contains basic information about a joint like its type and name. It can be obtained via
+/// [`get_joint_info()`](`crate::PhysicsClient::get_joint_info()`)
+/// # Example
+/// ```rust
+/// use rubullet::{PhysicsClient, UrdfOptions};
+/// use nalgebra::Isometry3;
+/// use rubullet::mode::Mode::Direct;
+/// use easy_error::Terminator;
+/// fn main() -> Result<(),Terminator> {
+///
+///     let mut client = PhysicsClient::connect(Direct)?;
+///     client.set_additional_search_path("../rubullet-sys/bullet3/libbullet3/data")?;
+///     client.set_additional_search_path(
+///         "../rubullet-sys/bullet3/libbullet3/examples/pybullet/gym/pybullet_data",
+///         )?;
+///     let panda_id = client.load_urdf("franka_panda/panda.urdf", UrdfOptions::default())?;
+///     let joint_info = client.get_joint_info(panda_id,4);
+///     assert_eq!("panda_joint5",joint_info.joint_name);
+///     Ok(())
+/// }
+/// ```
+/// # See also
+/// * [`JointState`](`crate::types::JointState`) - For information about the current state of the joint.
+pub struct JointInfo {
+    /// the same joint index as the input parameter
+    pub joint_index: i32,
+    /// the name of the joint, as specified in the URDF (or SDF etc) file
+    pub joint_name: String,
+    /// type of the joint, this also implies the number of position and velocity variables.
+    pub joint_type: JointType,
+    /// the first position index in the positional state variables for this body
+    pub q_index: i32,
+    /// the first velocity index in the velocity state variables for this body
+    pub u_index: i32,
+    /// reserved
+    pub flags: i32,
```

Comment:
:shrug: However I use it in the tests to verify that it has the same output as in pybullet.
these should be the possible values:
https://github.com/bulletphysics/bullet3/blame/875a6a4446c8b576125c96bc27421313ff2f2475/examples/SharedMemory/SharedMemoryPublic.h#L275

They were added 4 years ago and seem to have something to with VR.
I would say we should  implement it with the bitflags crate or make it #[doc(hidden)] or both

Reply:
We had flags before that we decided to make bitflags and hidden, right? Let's do that again to keep things consistent.

Reply:
We have one where we used bitflags. and two where we used Option<i32> and doc(hidden) because we did not know what their variants are. We could look them up and decide whether they are known and if they are "stable". Or we could keep the flag as i32 and make it hidden.

Reply:
Looking at this without the context of the others, I would say bitflags+hidden makes the most sense. Eventually we should probably be consistent, but I don't have enough of a mental model right now to really work that out.

(I'm commenting on these in between resetting the robot, so I'm not able to spend much time building that model between comments)

Reply:
done.